### ISSUE TITLE
Incompatible with pyopencl==2024.3

### ISSUE THREAD

**Issue by sablin39** [sablin39/pyrqa#1]
Get errors with `python pyrqa.test`:

'''
AttributeError: module 'pyopencl' has no attribute 'enqueue_write_buffer'

AttributeError: module 'pyopencl' has no attribute 'enqueue_read_buffer'
'''

**Comment by sablin39** [comment]
According to issue githubharald/CTCDecoder#5 and commit inducer/pyopencl@0a730d5e5be3df3229ea14c83dd2b6483bca843d, these two are deprecated and merged to enqueue_copy(queue, dest, src, **kwargs)[[source]](https://github.com/inducer/pyopencl/blob/main/pyopencl/__init__.py#L1768-L2066) .

**Comment by sablin39** [comment]
Seems resolved in 547b6e7

Conversion of `enqueue_write_buffer` is resolved with: 
```bash
find . -type f -name '*.py' -exec sed -i 's/cl\.enqueue_write_buffer/cl\.enqueue_copy/g' {} +
```

Conversion of `enqueue_read_buffer` needs to change order of arguments, so I use the following python script:
```python
import libcst as cst
import libcst.matchers as m
import sys
import argparse
from typing import Optional, List
import os
from tqdm import tqdm

class EnqueueReadBufferTransformer(cst.CSTTransformer):
    def leave_Call(
        self, original_node: cst.Call, updated_node: cst.Call
    ) -> Optional[cst.Call]:
        """
        Transforms cl.enqueue_read_buffer(...) to cl.enqueue_copy(...),
        swapping the second and third positional arguments.
        """
        # Check if the function is cl.enqueue_read_buffer
        if m.matches(
            updated_node.func,
            m.Attribute(
                value=m.Name("cl"),
                attr=m.Name("enqueue_read_buffer")
            )
        ):
            # Change function name to enqueue_copy
            new_func = updated_node.func.with_changes(attr=cst.Name("enqueue_copy"))

            # Extract arguments
            args = updated_node.args

            # Ensure there are at least three positional arguments to swap
            positional_args = [arg for arg in args if arg.keyword is None]
            if len(positional_args) >= 3:
                # Find indices of the second and third positional arguments
                # to swap them
                # Note: LibCST maintains all arguments (positional and keyword) in a list
                # We need to identify the positions of the second and third positional args
                pos_indices = [i for i, arg in enumerate(args) if arg.keyword is None]
                if len(pos_indices) >= 3:
                    second_pos = pos_indices[1]
                    third_pos = pos_indices[2]

                    # Swap the second and third positional arguments
                    new_args = list(args)
                    new_args[second_pos], new_args[third_pos] = new_args[third_pos], new_args[second_pos]

                    # Create a new Call node with updated arguments
                    return updated_node.with_changes(func=new_func, args=new_args)
            else:
                # Not enough positional arguments to swap; issue a warning
                print(f"Warning: cl.enqueue_read_buffer called with fewer than 3 positional arguments at line {original_node.lineno}")
                # Still change the function name
                return updated_node.with_changes(func=new_func)

        return updated_node

def replace_enqueue_read_buffer(file_path: str):
    """
    Reads a Python file, replaces cl.enqueue_read_buffer with cl.enqueue_copy
    (swapping the second and third arguments), and writes the changes back.
    """
    with open(file_path, 'r', encoding='utf-8') as file:
        source_code = file.read()

    # Parse the source code into a CST
    try:
        tree = cst.parse_module(source_code)
    except cst.ParserSyntaxError as e:
        print(f"SyntaxError while parsing {file_path}: {e}")
        return

    # Transform the CST
    transformer = EnqueueReadBufferTransformer()
    transformed_tree = tree.visit(transformer)

    # Convert the CST back to source code
    new_source_code = transformed_tree.code

    # Write the transformed code back to the file
    with open(file_path, 'w', encoding='utf-8') as file:
        file.write(new_source_code)

    print(f"Successfully updated {file_path}")

def list_py_files_recursive(folder_path):
    pyfiles=[]
    for root, dirs, files in os.walk(folder_path):
        for file in files:
            if file.endswith('.py'):
                # print(os.path.join(root, file))
                pyfiles.append(os.path.join(root, file))
    return pyfiles

def main():
    parser = argparse.ArgumentParser(description="Replace cl.enqueue_read_buffer with cl.enqueue_copy in a Python file while preserving formatting and comments.")
    parser.add_argument('file', help="Path to the Python file to be processed.")
    args =   @@parser.parse_args()

    files=list_py_files_recursive(args.file)
    pbar=tqdm(files)
    for file in pbar:
        pbar.set_description(f"Processing {file}")
        replace_enqueue_read_buffer(file)

if __name__ == "__main__":
    main()

```

**Comment by sablin39** [comment]
Completed with all tests passed in approximately 120s on Nvidia RTX3090
